modules/substrate/src/verifier.rs

```diff
@@ -0,0 +1,667 @@
+{
+	fn import_header(storage: &mut S, header: H) -> Result<(), ImportError> {
+		let best_finalized = storage.best_finalized_header();
+
+		if header.number() < best_finalized.number() {
+			return Err(ImportError::OldHeader);
+		}
+
+		if storage.header_exists(header.hash()) {
+			return Err(ImportError::HeaderAlreadyExists);
+		}
+
+		let parent_header = storage.get_header_by_hash(*header.parent_hash());
+		if parent_header.is_none() {
+			return Err(ImportError::MissingParent);
+		}
+
+		// We don't need the justification right away, but we should note it.
+		let requires_justification = find_scheduled_change(&header).is_some();
+		let is_finalized = false;
+		storage.write_header(&ImportedHeader::new(header, requires_justification, is_finalized));
+
+		Ok(())
+	}
+
+	fn verify_finality(storage: &mut S, hash: H::Hash, proof: FinalityProof) -> Result<(), ImportError> {
+		// Make sure that we've previously imported this header
+		let header = storage.get_header_by_hash(hash).ok_or(ImportError::UnknownHeader)?;
+
+		let current_authority_set = storage.current_authority_set();
+		let is_finalized = prove_finality(&header, &current_authority_set, proof);
+		if !is_finalized {
+			return Err(ImportError::UnfinalizedHeader);
+		}
+
+		let last_finalized = storage.best_finalized_header();
+		let mut finalized_headers = if let Some(ancestors) = are_ancestors(storage, last_finalized, header.clone()) {
+			// Skip header we're trying to finalize since we know it `requires_justification`
+			let requires_justification = ancestors.iter().skip(1).find(|h| h.requires_justification);
+
+			// This means that we're trying to import a justification for the child
+			// of a header which is still missing a justification. We must reject
+			// this header.
+			if requires_justification.is_some() {
+				return Err(ImportError::PrematureJustification);
+			}
+
+			let current_set_id = current_authority_set.set_id;
+			update_authority_set(storage, &header.header, current_set_id)?;
+			ancestors
+		} else {
+			return Err(ImportError::AncestryCheckFailed);
+		};
+
+		for header in finalized_headers.iter_mut() {
+			// TODO: Maybe mutate() storage?
+			header.is_finalized = true;
+			header.requires_justification = false;
+			storage.write_header(header);
+		}
+
+		let best_finalized = finalized_headers
+			.last()
+			.expect("We just iterated through these headers, therefore the last header must exist");
+		storage.update_best_finalized((*best_finalized).hash());
+
+		Ok(())
+	}
+}
+
+// Returns the lineage of headers from (ancestor, child]
+fn are_ancestors<S, H>(
+	storage: &S,
+	ancestor: ImportedHeader<H>,
+	child: ImportedHeader<H>,
+) -> Option<Vec<ImportedHeader<H>>>
+where
+	S: BridgeStorage<Header = H>,
+	H: HeaderT,
+{
+	let mut ancestors = vec![];
+	let mut current_header = child;
+
+	while ancestor.hash() != current_header.hash() {
+		// We've gotten to the same height and we're not related
+		if ancestor.number() == current_header.number() {
```

Comment:
```suggestion
		if ancestor.number() >= current_header.number() {
```
Early exit if `ancestor.number() > child.number()`